lib/df-detection.c

```diff
@@ -0,0 +1,242 @@
+#include <linux/debugfs.h>
+#include <linux/df-detection.h>
+#include <linux/fs.h>
+#include <linux/init.h>
+#include <linux/kernel.h>
+#include <linux/random.h>
+#include <linux/sched.h>
+#include <linux/slab.h>
+#include <linux/stacktrace.h>
+#include <linux/sys.h>
+#include <linux/types.h>
+
+static int is_intersect(struct dfetch_address_range a,
+			struct dfetch_address_range b, void *kernel_addr)
+{
+	void *a_end = (void *)((char *)a.start_address + a.len);
+	void *b_end = (void *)((char *)b.start_address + b.len);
+	size_t len;
+	void *addr = kernel_addr;
+
+	if (a.start_address <= b.start_address && a_end > b.start_address) {
+		len = (char *)(a_end > b_end ? b_end : a_end) -
+		      (char *)b.start_address;
+	} else if (b.start_address <= a.start_address &&
+		   b_end > a.start_address) {
+		unsigned long diff =
+		    (char *)a.start_address - (char *)b.start_address;
+
+		len = (char *)(a_end > b_end ? b_end : a_end) -
+		      (char *)a.start_address;
+		addr = ((char *)kernel_addr + diff);
+	} else
+		return 0;
+	get_random_bytes(addr, len);
+	pr_err("Random injection in %px, Length %lu \n", addr, len);
+	return 1;
+}
+
+static void detect_intersection(void *kernel_addr)
+{
+	int i;
+
+	for (i = 0; i < current->num_read; i++) {
+		if (!is_intersect(current->dfetch_addresses[i],
+				  current->dfetch_addresses[current->num_read],
+				  kernel_addr))
+			continue;
+		if (current->dfetch_index >= current->dfetch_size &&
+		    current->dfetch_size <
+			DFETCH_MAX_RECORDS * DFETCH_MAX_RECORDS) {
+			struct dfetch_pair *temp =
+			    (struct dfetch_pair *)krealloc(
+				current->dfetch_pairs,
+				current->dfetch_size * 2 *
+				    sizeof(struct dfetch_pair),
+				GFP_KERNEL);
+			current->dfetch_pairs =
+			    temp ? temp : current->dfetch_pairs;
+			current->dfetch_size = temp ? current->dfetch_size * 2
+						    : current->dfetch_size;
+		}
+		if (current->dfetch_index < current->dfetch_size) {
+			current->dfetch_pairs[current->dfetch_index].first = i;
+			current->dfetch_pairs[current->dfetch_index].second =
+			    current->num_read;
+			current->dfetch_index++;
+		}
+	}
+}
+
+static void report(void)
+{
+	int i;
+	unsigned long *first_entries;
```

Comment:
It'll be more compact if you declare first_entries and second_entries together on the same line.
Same for xxx_nr_entries.

Reply:
I even think replacing every pair of variables with an array will make the code more compact.